shell/platform/android/android_external_texture_gl.cc

```diff
@@ -64,23 +64,44 @@ void AndroidExternalTextureGL::Paint(SkCanvas& canvas,
       kPremul_SkAlphaType, nullptr);
   if (image) {
     SkAutoCanvasRestore autoRestore(&canvas, true);
-    canvas.translate(bounds.x(), bounds.y());
-    canvas.scale(bounds.width(), bounds.height());
-    if (!transform.isIdentity()) {
-      SkMatrix transformAroundCenter(transform);
 
-      transformAroundCenter.preTranslate(-0.5, -0.5);
-      transformAroundCenter.postScale(1, -1);
-      transformAroundCenter.postTranslate(0.5, 0.5);
-      canvas.concat(transformAroundCenter);
+    // The incoming texture is vertically flipped, so we flip it
+    // back. OpenGL's coordinate system has Positive Y equivalent to up, while
+    // Skia's coordinate system has Negative Y equvalent to up.
+    canvas.translate(bounds.x(), bounds.y() + bounds.height());
+    canvas.scale(bounds.width(), -bounds.height());
+
+    if (!transform.isIdentity()) {
+      sk_sp<SkShader> shader = image->makeShader(
+          SkTileMode::kRepeat, SkTileMode::kRepeat, sampling, transform);
+
+      SkPaint paintWithShader;
+      if (paint) {
+        paintWithShader = *paint;
+      }
+      paintWithShader.setShader(shader);
+      canvas.drawRect(SkRect::MakeWH(1, 1), paintWithShader);
+    } else {
+      canvas.drawImage(image, 0, 0, sampling, paint);
```

Comment:
Won't this draw the image scaled? Shouldn't it be a drawImageRect? Was testing done outside the context of a DPI scale?

Reply:
The GrBackendTexture created above has a size of 1x1 since the texture's actual  size is dictated by Android

Reply:
We're already using `canvas.drawImage` in existing code, and texture layers always get drawn underneath a layer that applies the DPI scale....